src/Condition.js

```diff
@@ -0,0 +1,303 @@
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+// To avoid forward referencing the transformer function
+let transform;
+
+/**
+ * Determines how to transform children configuration based on the affix type.
+ */
+const configMapper = {
+  prefix: cfg => transform(cfg),
+  infix: cfg => cfg.map(child => transform(child)),
+};
+
+/**
+ * Determines how to compose VCL based on the affix type.
+ */
+const vclComposer = {
+  prefix: op => item => `${op}(${item.toVCL()})`,
+  infix: op => items => `(${items.map(item => item.toVCL()).join(op)})`,
+};
+
+/**
+ * Boolean conditions
+ */
+const booleanMap = {
+  or: {
+    mapper: configMapper.infix,
+    vcl: vclComposer.infix(' || '),
+    express: (items, req) => items.reduce((result, item) => result || item.evaluate(req), false),
+  },
+  and: {
+    mapper: configMapper.infix,
+    vcl: vclComposer.infix(' && '),
+    express: (items, req) => items.reduce((result, item) => result && item.evaluate(req), true),
+  },
+  not: {
+    mapper: configMapper.prefix,
+    vcl: vclComposer.prefix(' !'),
+    express: (item, req) => !item.evaluate(req),
+  },
+};
+
+class BooleanCondition {
+  constructor(entry, cfg) {
+    this._entry = entry;
+    this._items = this._entry.mapper(cfg);
+  }
+
+  toVCL() {
+    return this._entry.vcl(this._items, this._entry.op);
+  }
+
+  evaluate(req) {
+    return this._entry.express(this._items, req);
+  }
+}
+
+/**
+ * PropertyCondition
+ */
+class PropertyCondition {
+  constructor(prop, op, value, name) {
+    if (op && prop.allowed_ops.indexOf(op) === -1) {
+      throw new Error(`Property ${name} does not support operation: ${op}`);
+    }
+
+    this._prop = prop;
+    this._op = op;
+    this._value = value;
+    this._name = name;
+  }
+
+  toVCL() {
+    const quote = this._prop.type === 'string' ? '"' : '';
+    let value = this._value;
+    let op = this._op;
+
+    if (quote === '"' && !op) {
+      // substring-start
+      value = `^${value}`;
+      op = '~';
+    }
+    if (!op || op === '=') {
+      // operand defaults to equal
+      op = '==';
+    }
+    const vcl = typeof this._prop.vcl === 'function' ? this._prop.vcl(this) : this._prop.vcl;
+    return `${vcl} ${op} ${quote}${value}${quote}`;
+  }
+
+  evaluate(req) {
+    if (!this._prop.express) {
+      return true;
+    }
+    const actual = this._prop.express(req, this);
+    if (!actual) {
+      return false;
+    }
+
+    const value = this._value;
+    const { type } = this._prop;
+
+    switch (this._op) {
+      case '=':
+        return actual === value;
+      case '~':
+        return !!actual.match(value);
+      case '<':
+        return actual < value;
+      case '>':
+        return actual > value;
+      default:
+        return type === 'string' ? actual.startsWith(value) : actual === value;
```

Comment:
sorry to bother again. but this is not perfect for urls and url paths. because we currently use path-aware startsWith:

assume you have an urls:

```yaml
condition:
  - url: https://adobe.com/doc
```

then you want to match `https://adobe.com/doc` and `https://adobe.com/doc/` and `https://adobe.com/doc/index.html` but **NOT** `https://adobe.com/documentation/index.html`

it can be written as regexp:

```yaml
condition:
  - url~: https://adobe.com/doc(/.*)?
```

However, this is a very common case, and we should either implement it this way, other document it and provide all examples accordingly. But, it will complicate the baseURL computation that we need to provide (in fastly and in function)

alternatively, we can also mandate that people write:

```yaml
condition:
  - url: https://adobe.com/doc/
redirects:
  - from: /doc 
    to: /doc/
```


@trieloff WDYT?

Reply:
I agree, the `url.path: /doc` should match:

- `/doc`
- `/doc/umentation`
- `/doc/index.html`

but not

- `/documentation/`
- `/doctorwho`
- `/doctorstrange`

So what I called "substring match" is actually a "parent directory match".

Forcing people to use RegExes (which come with other downsides) or redirects (slower – and you still have the question of the condition) is suboptimal.

Reply:
Right, but this "parent directory match" just applies to the "missing" operator, not the other ones, `=` or `~`, doesn't it?

Reply:
correct.